### ISSUE TITLE
get stuck if a popup filter causes an error

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6999]
**Describe the bug**

We can get stuck if a popup filter causes an error.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        let g:id = popup_create('', #{filter: 'invalid'})
        call timer_start(0, {-> win_execute(g:id, 'norm! zz')})
    EOF
    )

`E994` and `E117` are raised:

    Error detected while processing function <lambda>1:
    line    1:
    E994: Not allowed in a popup window
    Error detected while processing function <lambda>1:
    line    1:
    E117: Unknown function: invalid

This is expected.

The errors are raised continuously: this is not expected.

**Expected behavior**

The errors are raised only once.

**Environment**

 - Vim version: 8.2 Included patches: 1-1724
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(360)

**Additional context**

We can stop the errors by pressing `C-c`.

---

This issue looks similar to #6902 which was fixed by [8.2.1636](https://github.com/vim/vim/releases/tag/v8.2.1636).

**Comment by lacygoill** [comment]
> This is expected.

Actually, I'm not sure `E994` is expected.  `norm! zz` works in a popup window when no error is raised.  But I guess `E994` is here caused by `E117`.

**Comment by brammool** [comment]
> **Describe the bug**
> 
> We can get stuck if a popup filter causes an error.
> 
> **To Reproduce**
> 
> Run this shell command:
> 
>     vim -Nu NONE -S <(cat <<'EOF'
>         let g:id = popup_create('', #{filter: 'invalid'})
>         call timer_start(0, {-> win_execute(g:id, 'norm! zz')})
>     EOF
>     )
> 
> `E994` and `E117` are raised:
> 
>     Error detected while processing function <lambda>1:
>     line    1:
>     E994: Not allowed in a popup window
>     Error detected while processing function <lambda>1:
>     line    1:
>     E117: Unknown function: invalid
> 
> This is expected.
> 
> The errors are raised continuously: this is not expected.
> 
> **Expected behavior**
> 
> The errors are raised only once.
> 
> **Environment**
> 
>  - Vim version: 8.2 Included patches: 1-1724
>  - OS: Ubuntu 16.04.7 LTS
>  - Terminal: xterm(360)
> 
> **Additional context**
> 
> We can stop the errors by pressing `C-c`.

I can fix it, but haven't yet found a test to catch the problem.
Can you think of a way this would make a test fail, while the tests
passes once the problem is fixed?

-- 
BEDEVERE: Wait.  Wait ... tell me, what also floats on water?
ALL:      Bread?  No, no, no.  Apples .... gravy ... very small rocks ...
ARTHUR:   A duck.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
This is above my level, but I would naively copy and adapt the test added in 8.2.1636.  The only difference is the command which triggers the issue.  In #6902, the issue is triggered by an interactive normal command.  Here, the issue is triggered by a non-interactive normal command run from a timer's callback.  I guess we would need a `term_wait()` between the sourcing of the script and `VerifyScreenDump()`.  I came up with this test which currently fails:
```diff
diff --git a/src/testdir/dumps/Test_popupwin_wrong_name.dump b/src/testdir/dumps/Test_popupwin_wrong_name_1.dump
similarity index 100%
rename from src/testdir/dumps/Test_popupwin_wrong_name.dump
rename to src/testdir/dumps/Test_popupwin_wrong_name_1.dump
diff --git a/src/testdir/dumps/Test_popupwin_wrong_name_2.dump b/src/testdir/dumps/Test_popupwin_wrong_name_2.dump
new file mode 100644
index 000000000..1d7b16428
--- /dev/null
+++ b/src/testdir/dumps/Test_popupwin_wrong_name_2.dump
@@ -0,0 +1,10 @@
+|E+0#ffffff16#e000002|9@1|4|:| |N|o|t| |a|l@1|o|w|e|d| |i|n| |a| |p|o|p|u|p| |w|i|n|d|o|w| +0#0000000#ffffff0@39
+|P+0#00e0003&|r|e|s@1| |E|N|T|E|R| |o|r| |t|y|p|e| |c|o|m@1|a|n|d| |t|o| |c|o|n|t|i|n|u|e| +0#0000000&@35
+|E+0#ffffff16#e000002|r@1|o|r| |d|e|t|e|c|t|e|d| |w|h|i|l|e| |p|r|o|c|e|s@1|i|n|g| |f|u|n|c|t|i|o|n| |<|l|a|m|b|d|a|>|1|:| +0#0000000#ffffff0@23
+|l+0#af5f00255&|i|n|e| @3|1|:| +0#0000000&@64
+|E+0#ffffff16#e000002|1@1|7|:| |U|n|k|n|o|w|n| |f|u|n|c|t|i|o|n|:| |N|o|S|u|c|h|F|u|n|c| +0#0000000#ffffff0@40
+|P+0#00e0003&|r|e|s@1| |E|N|T|E|R| |o|r| |t|y|p|e| |c|o|m@1|a|n|d| |t|o| |c|o|n|t|i|n|u|e| +0#0000000&@35
+|E+0#ffffff16#e000002|r@1|o|r| |d|e|t|e|c|t|e|d| |w|h|i|l|e| |p|r|o|c|e|s@1|i|n|g| |f|u|n|c|t|i|o|n| |<|l|a|m|b|d|a|>|1|:| +0#0000000#ffffff0@23
+|l+0#af5f00255&|i|n|e| @3|1|:| +0#0000000&@64
+|E+0#ffffff16#e000002|9@1|4|:| |N|o|t| |a|l@1|o|w|e|d| |i|n| |a| |p|o|p|u|p| |w|i|n|d|o|w| +0#0000000#ffffff0@39
+|P+0#00e0003&|r|e|s@1| |E|N|T|E|R| |o|r| |t|y|p|e| |c|o|m@1|a|n|d| |t|o| |c|o|n|t|i|n|u|e> +0#0000000&@35
diff --git a/src/testdir/test_popupwin.vim b/src/testdir/test_popupwin.vim
index 333e9a284..e8218068e 100644
--- a/src/testdir/test_popupwin.vim
+++ b/src/testdir/test_popupwin.vim
@@ -3543,9 +3543,24 @@ func Test_popupwin_filter_close_wrong_name()
   let buf = RunVimInTerminal('-S XtestPopupWrongName', #{rows: 10})
 
   call term_sendkeys(buf, "j")
-  call VerifyScreenDump(buf, 'Test_popupwin_wrong_name', {})
+  call VerifyScreenDump(buf, 'Test_popupwin_wrong_name_1', {})
 
   call StopVimInTerminal(buf)
+
+  let lines =<< trim END
+      let g:winid = popup_create('one two three...', {'filter': 'NoSuchFunc'})
+      call timer_start(0, {-> win_execute(g:winid, 'norm! zz')})
+  END
+  call writefile(lines, 'XtestPopupWrongName')
+
+  let buf = RunVimInTerminal('-S XtestPopupWrongName', #{rows: 10})
+  call TermWait(buf)
+
+  call VerifyScreenDump(buf, 'Test_popupwin_wrong_name_2', {})
+
+  call StopVimInTerminal(buf)
+  unlet! g:winid
+
   call delete('XtestPopupWrongName')
 endfunc
 
```
If this doesn't work, I would need more time to study how tests work when run in a Vim terminal.

**Comment by brammool** [comment]
> This is above my level, but I would naively copy and adapt the test
> added in 8.2.1636.  The only difference is the command which triggers
> the issue.  In #6902, the issue is triggered by an interactive normal
> command.  Here, the issue is triggered by a non-interactive normal
> command run from a timer's callback.  I guess we would need a
> `term_wait()` between the sourcing of the script and
> `VerifyScreenDump()`.  I came up with this test which currently fails:

I was trying to do it without a screendump, but I suppose that's too
tricky, since there will always be errors.  I'll do something like you
suggested.  Hopefully it won't be flaky.

-- 
Never under any circumstances take a sleeping pill
and a laxative on the same night.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
Fixed by 8.2.1729.

**Comment by lacygoill** [comment]
Although, it seems the issue persists if both the function name and the command executed in the timer's callback are invalid:
```vim
let g:id = popup_create('', #{filter: 'invalid'})
call timer_start(0, {-> win_execute(g:id, 'invalid')})
```
And this time, not even `C-c` can stop the errors.
